shared/src/main/java/org/odk/collect/shared/PathUtils.kt

```diff
@@ -11,7 +11,13 @@ object PathUtils {
 
     @JvmStatic
     fun getAbsoluteFilePath(dirPath: String, filePath: String): String {
-        return if (filePath.startsWith(dirPath)) filePath else dirPath + File.separator + filePath
+        val absolutePath = if (filePath.startsWith(dirPath)) filePath else dirPath + File.separator + filePath
+
+        if (File(absolutePath).canonicalPath.startsWith(dirPath)) {
+            return absolutePath
+        } else {
+            throw SecurityException()
```

Comment:
I was initially considering checking this only during form upload, but we can proceed with this approach. However, I believe we should at least include another test to confirm that the exception is triggered during form upload. Right now, we know this method is used in that context, but it's not guaranteed. There's a risk that we might refactor the code and overlook this exception that should be thrown.